### ISSUE TITLE
Pass in API key/secret/token to constructor (remove the need for .env)

### ISSUE THREAD

**Issue by wxni** [slimandslam/schwab-client-js#1]
Would it be possible to add the ability to manually pass in the API key/secret/token to avoid the need for a `.env` file.

I'm currently using a secrets manager, so this would require a redeployment of my application without any code changes made.

My initial thought was a new constructor overload for `SchwabAPIclient` that accepts the three parameters: i.e.`SchwabAPIclient(appKey, appSecret, appRefresh);`

**Comment by slimandslam** [comment]
You can set them as environment variables instead of using a .env file. Is that a possible solution for you?

**Comment by wxni** [comment]
I'm currently using Infisical, with a custom implementation to update the refresh token, on Lambda.

At runtime I pull: key, secret, and token so I would want to avoid anything that would require weekly changes to the deployment or function.

Would a constructor with those values being optional parameters be a big change?

**Comment by slimandslam** [comment]
Can you just add the vars to your Lambda environment like this?
```
import { InfisicalClient } from "infisical";
import AWS from "aws-sdk";
const client = new InfisicalClient({token: INFISICAL_TOKEN  });
const secrets = await client.getAllSecrets();
const myenv = {
        SCHWAB_APP_KEY: secrets["SCHWAB_APP_KEY"],
        SCHWAB_SECRET: secrets["SCHWAB_SECRET"],
        SCHWAB_REFRESH_TOKEN: secrets["SCHWAB_REFRESH_TOKEN"],
    };
const lambda = new AWS.Lambda({ region: "us-east-1" });
const theLambdaName = "my-lambda-function";
await lambda.updateFunctionConfiguration({
    FunctionName: theLambdaName,
    Environment: { Variables: myenv }
  }).promise();
```
This code will overwrite any other vars in your Lambda env, so you'll have to modify it a bit to merge with existing env vars if that's an issue.

**Comment by wxni** [comment]
Even thought it's fairly bad practice to modify environment variables at runtime, I did try it and it doesn't seem to work with my current deployment structure nor existing environment variables.

Would something like this suffice:
```
class SchwabAPIclient {
  // adding optional parameters to the constructor
  constructor(appKey: string = "", appSecret: string = "", appRefresh: string = "") {
    const credentials: Credentials = {
      appKey: process.env.SCHWAB_APP_KEY || appKey,
      appSecret: process.env.SCHWAB_SECRET || appSecret,
      appRefresh: process.env.SCHWAB_REFRESH_TOKEN || appRefresh, // expires every seven days
      access_token: "", // expires every 30 minutes
      access_exp: Math.floor(Date.now() / 1000) + 29 * 60, // expiration is 29 minutes in the future
    };
...
  }
```
  
Then when any of the three classes: MarketApiClient, TradingApiClient, StreamingApiClient are called they can just be called like so:
```
...
  const secrets = await infisicalClient.secrets().listSecrets({
    environment: "prod",
    projectId: projectId,
  });

const mktclient = new MarketApiClient(secrets["SCHWAB_APP_KEY"], secrets["SCHWAB_SECRET"], secrets["SCHWAB_REFRESH_TOKEN"]);
...
```

These would all be optional parameters, so it shouldn't require any modification outside of the class structure if I'm not mistaken and no other modifications should be necessary?

I really appreciate your time and thoughtful responses as well, please keep up the great work since this repo is a lot better than anything I've seen for JS/TS with Schwab's API!

**Comment by slimandslam** [comment]
Yes, that would likely work. I'm working on some other changes right now, but I'll get to it.

Can I get some details of your setup? How are you triggering the Lambda function? Is Infisical
local or on a managed server somewhere?

I'm curious as to why changing the environment vars doesn't work for you. If you 
merge the new refresh key using step #1 below, and then immediately  publish a new version with step #2, 
you are guaranteed to have the new env be $LATEST (I assume that's the version you invoke)
 and all further Lambda invocations should use it.
1. await updateFunctionConfiguration()
2. await publishVersion()

**Comment by wxni** [comment]
I really appreciate that.

- Serverless to deploy
- Cloudwatch to trigger the function
- Infisical managed server to handle Schwab specific information

The reason this wont work for my use case is because this would require infrastructure as code to be implemented into the application layer as opposed to being managed by deployments i.e. `serverless.yml`. Quite a few of my functions are duplicates of one another with the only differences being in the deployment. I would then have to create a new function (with identical code) specifically for environment variable changes.

Above all, I would hate to be a burden, it seems like a very small change, would you like me to create a PR adding those optional parameters? I can do my best to update the README and create an example file if you would like.

If you would like you can email me (from my bio) and we can discuss further on Discord etc and I can give more details of my setup.
